### DISCUSSION TITLE
Python websocket-server issue

### DISCUSSION THREAD

**Discussion by ceejayemm** [rern/rAudio#1473] [Fixed Issues]
@rern 
I originally reported something similar to this as #1411 but it is now happening with 20231022..  When I reboot the RPi I am seeing, vai 'journalctl -xe'

ct 27 17:28:06 RAConservatory websocket-server.py[290]: connection handler failed
Oct 27 17:28:06 RAConservatory websocket-server.py[290]: Traceback (most recent call last):
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 968, in transfer_data
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     message = await self.read_message()
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:               ^^^^^^^^^^^^^^^^^^^^^^^^^
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 1038, in read_message
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     frame = await self.read_data_frame(max_size=self.max_size)
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 1113, in read_data_frame
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     frame = await self.read_frame(max_size)
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 1170, in read_frame
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     frame = await Frame.read(
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:             ^^^^^^^^^^^^^^^^^
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/framing.py", line 69, in read
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     data = await reader(2)
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:            ^^^^^^^^^^^^^^^
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/usr/lib/python3.11/asyncio/streams.py", line 733, in readexactly
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     await self._wait_for_data('readexactly')
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/usr/lib/python3.11/asyncio/streams.py", line 526, in _wait_for_data
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     await self._waiter
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/usr/lib/python3.11/asyncio/selector_events.py", line 995, in _read_ready__data_received
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     data = self._sock.recv(self.max_size)
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Oct 27 17:28:06 RAConservatory websocket-server.py[290]: ConnectionResetError: [Errno 104] Connection reset by peer
Oct 27 17:28:06 RAConservatory websocket-server.py[290]: The above exception was the direct cause of the following exception:
Oct 27 17:28:06 RAConservatory websocket-server.py[290]: Traceback (most recent call last):
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/server.py", line 236, in handler
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     await self.ws_handler(self)
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/server.py", line 1175, in _ws_handler
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     return await cast(
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:            ^^^^^^^^^^^
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/srv/http/bash/websocket-server.py", line 10, in cmd
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     async for args in websocket:
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 497, in __aiter__
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     yield await self.recv()
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:           ^^^^^^^^^^^^^^^^^
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 568, in recv
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     await self.ensure_open()
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 944, in ensure_open
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     raise self.connection_closed_exc()
Oct 27 17:28:06 RAConservatory websocket-server.py[290]: websockets.exceptions.ConnectionClosedError: no close frame received or sent
Oct 27 17:28:06 RAConservatory websocket-server.py[290]: connection handler failed
Oct 27 17:28:06 RAConservatory websocket-server.py[290]: Traceback (most recent call last):
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 968, in transfer_data
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     message = await self.read_message()
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:               ^^^^^^^^^^^^^^^^^^^^^^^^^
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 1038, in read_message
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     frame = await self.read_data_frame(max_size=self.max_size)
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 1113, in read_data_frame
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     frame = await self.read_frame(max_size)
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 1170, in read_frame
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     frame = await Frame.read(
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:             ^^^^^^^^^^^^^^^^^
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/framing.py", line 69, in read
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     data = await reader(2)
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:            ^^^^^^^^^^^^^^^
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/usr/lib/python3.11/asyncio/streams.py", line 733, in readexactly
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     await self._wait_for_data('readexactly')
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/usr/lib/python3.11/asyncio/streams.py", line 526, in _wait_for_data
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     await self._waiter
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/usr/lib/python3.11/asyncio/selector_events.py", line 995, in _read_ready__data_received
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     data = self._sock.recv(self.max_size)
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Oct 27 17:28:06 RAConservatory websocket-server.py[290]: ConnectionResetError: [Errno 104] Connection reset by peer
Oct 27 17:28:06 RAConservatory websocket-server.py[290]: The above exception was the direct cause of the following exception:
Oct 27 17:28:06 RAConservatory websocket-server.py[290]: Traceback (most recent call last):
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/server.py", line 236, in handler
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     await self.ws_handler(self)
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/server.py", line 1175, in _ws_handler
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     return await cast(
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:            ^^^^^^^^^^^
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/srv/http/bash/websocket-server.py", line 10, in cmd
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     async for args in websocket:
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 497, in __aiter__
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     yield await self.recv()
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:           ^^^^^^^^^^^^^^^^^
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 568, in recv
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     await self.ensure_open()
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 944, in ensure_open
Oct 27 17:28:06 RAConservatory websocket-server.py[290]:     raise self.connection_closed_exc()
Oct 27 17:28:06 RAConservatory websocket-server.py[290]: websockets.exceptions.ConnectionClosedError: no close frame received or sent

It only seems to happen on a boot / reboot and doesn't seem to subsequently reoccur. Is this something which needs to be sorted out ?

Chris

**Comment by ceejayemm** [comment]
@rern 
This is still happening with 20231105

Chris

**Comment by ceejayemm** [comment]
@rern 
Still happening with i20231022 updated to 20231111.  If this something we don't need to worry about then please let me know and I will close the issue.

Chris

**Comment by ceejayemm** [comment]
@rern 
Still happening with 20231128 updated to latest Update branch (as at 0930 GMT 0/12/23). It seems to be related to adding tracks to the Playlist and also using the Random option to randomise the playlist. Using 'journalctl -f' during sequence (starting with empty playlist) Select Artist / Add (all tracks) / View Playlist page / Randomize playlist / Refresh Playlist, I get the following output (which can be repeated as necessary):

**Starting with empty Playlist**

rAudio-1:~  #  journalctl -f

Dec 09 09:39:45 rAudio-1 sudo[65533]:     http : PWD=/srv/http ; USER=root ; COMMAND=/srv/http/bash/cmd.sh 'mpcaddfind#012artist#012Chris De Burgh#012CMD MODE STRING'
Dec 09 09:39:45 rAudio-1 sudo[65533]: pam_unix(sudo:session): session opened for user root(uid=0) by (uid=33)
Dec 09 09:39:45 rAudio-1 websocket-server.py[3996]: connection handler failed
Dec 09 09:39:45 rAudio-1 websocket-server.py[3996]: Traceback (most recent call last):
Dec 09 09:39:45 rAudio-1 websocket-server.py[3996]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/server.py", line 236, in handler
Dec 09 09:39:45 rAudio-1 websocket-server.py[3996]:     await self.ws_handler(self)
Dec 09 09:39:45 rAudio-1 websocket-server.py[3996]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/server.py", line 1175, in _ws_handler
Dec 09 09:39:45 rAudio-1 websocket-server.py[3996]:     return await cast(
Dec 09 09:39:45 rAudio-1 websocket-server.py[3996]:            ^^^^^^^^^^^
Dec 09 09:39:45 rAudio-1 websocket-server.py[3996]:   File "/srv/http/bash/websocket-server.py", line 10, in cmd
Dec 09 09:39:45 rAudio-1 websocket-server.py[3996]:     async for args in websocket:
Dec 09 09:39:45 rAudio-1 websocket-server.py[3996]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 497, in __aiter__
Dec 09 09:39:45 rAudio-1 websocket-server.py[3996]:     yield await self.recv()
Dec 09 09:39:45 rAudio-1 websocket-server.py[3996]:           ^^^^^^^^^^^^^^^^^
Dec 09 09:39:45 rAudio-1 websocket-server.py[3996]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 568, in recv
Dec 09 09:39:45 rAudio-1 websocket-server.py[3996]:     await self.ensure_open()
Dec 09 09:39:45 rAudio-1 websocket-server.py[3996]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 953, in ensure_open
Dec 09 09:39:45 rAudio-1 websocket-server.py[3996]:     raise self.connection_closed_exc()
Dec 09 09:39:45 rAudio-1 websocket-server.py[3996]: websockets.exceptions.ConnectionClosedError: received 1005 (no status code [internal]); then sent 1005 (no status code [internal])
Dec 09 09:39:45 rAudio-1 sudo[65533]: pam_unix(sudo:session): session closed for user root
Dec 09 09:39:45 rAudio-1 sudo[65547]:     http : PWD=/srv/http ; USER=root ; COMMAND=/srv/http/bash/status.sh
Dec 09 09:39:45 rAudio-1 sudo[65547]: pam_unix(sudo:session): session opened for user root(uid=0) by (uid=33)
Dec 09 09:39:49 rAudio-1 sudo[65547]: pam_unix(sudo:session): session closed for user root

**View resulting Playlist**

Dec 09 09:40:05 rAudio-1 nginx[454]: 2023/12/09 09:40:05 [error] 454#454: *603 FastCGI sent in stderr: "PHP message: PHP Warning:  Undefined variable $list in /srv/http/mpdplaylist.php on line 55; PHP message: PHP Warning:  Trying to access array offset on value of type null in /srv/http/mpdplaylist.php on line 55" while reading response header from upstream, client: 192.168.86.50, server: localhost, request: "POST /mpdplaylist.php HTTP/1.1", upstream: "fastcgi://unix:/run/php-fpm/php-fpm.sock:", host: "raudio-1", referrer: "http://raudio-1/"

**Randomize**

Dec 09 09:40:20 rAudio-1 nginx[454]: 2023/12/09 09:40:20 [error] 454#454: *603 FastCGI sent in stderr: "PHP message: PHP Warning:  Undefined variable $list in /srv/http/mpdplaylist.php on line 55; PHP message: PHP Warning:  Trying to access array offset on value of type null in /srv/http/mpdplaylist.php on line 55" while reading response header from upstream, client: 192.168.86.50, server: localhost, request: "POST /mpdplaylist.php HTTP/1.1", upstream: "fastcgi://unix:/run/php-fpm/php-fpm.sock:", host: "raudio-1", referrer: "http://raudio-1/"
Dec 09 09:40:29 rAudio-1 sudo[66232]:     http : PWD=/srv/http ; USER=root ; COMMAND=/srv/http/bash/cmd.sh mpcshuffle
Dec 09 09:40:29 rAudio-1 sudo[66232]: pam_unix(sudo:session): session opened for user root(uid=0) by (uid=33)
Dec 09 09:40:30 rAudio-1 sudo[66232]: pam_unix(sudo:session): session closed for user root
Dec 09 09:40:30 rAudio-1 websocket-server.py[3996]: connection handler failed
Dec 09 09:40:30 rAudio-1 websocket-server.py[3996]: Traceback (most recent call last):
Dec 09 09:40:30 rAudio-1 websocket-server.py[3996]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/server.py", line 236, in handler
Dec 09 09:40:30 rAudio-1 websocket-server.py[3996]:     await self.ws_handler(self)
Dec 09 09:40:30 rAudio-1 websocket-server.py[3996]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/server.py", line 1175, in _ws_handler
Dec 09 09:40:30 rAudio-1 websocket-server.py[3996]:     return await cast(
Dec 09 09:40:30 rAudio-1 websocket-server.py[3996]:            ^^^^^^^^^^^
Dec 09 09:40:30 rAudio-1 websocket-server.py[3996]:   File "/srv/http/bash/websocket-server.py", line 10, in cmd
Dec 09 09:40:30 rAudio-1 websocket-server.py[3996]:     async for args in websocket:
Dec 09 09:40:30 rAudio-1 websocket-server.py[3996]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 497, in __aiter__
Dec 09 09:40:30 rAudio-1 websocket-server.py[3996]:     yield await self.recv()
Dec 09 09:40:30 rAudio-1 websocket-server.py[3996]:           ^^^^^^^^^^^^^^^^^
Dec 09 09:40:30 rAudio-1 websocket-server.py[3996]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 568, in recv
Dec 09 09:40:30 rAudio-1 websocket-server.py[3996]:     await self.ensure_open()
Dec 09 09:40:30 rAudio-1 websocket-server.py[3996]:   File "/usr/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 953, in ensure_open
Dec 09 09:40:30 rAudio-1 websocket-server.py[3996]:     raise self.connection_closed_exc()
Dec 09 09:40:30 rAudio-1 websocket-server.py[3996]: websockets.exceptions.ConnectionClosedError: received 1005 (no status code [internal]); then sent 1005 (no status code [internal])
Dec 09 09:40:30 rAudio-1 sudo[66245]:     http : PWD=/srv/http ; USER=root ; COMMAND=/srv/http/bash/status.sh
Dec 09 09:40:30 rAudio-1 sudo[66245]: pam_unix(sudo:session): session opened for user root(uid=0) by (uid=33)
Dec 09 09:40:34 rAudio-1 sudo[66245]: pam_unix(sudo:session): session closed for user root

**Refresh Playlist page** 

Dec 09 09:40:48 rAudio-1 nginx[454]: 2023/12/09 09:40:48 [error] 454#454: *603 FastCGI sent in stderr: "PHP message: PHP Warning:  Undefined variable $list in /srv/http/mpdplaylist.php on line 55; PHP message: PHP Warning:  Trying to access array offset on value of type null in /srv/http/mpdplaylist.php on line 55" while reading response header from upstream, client: 192.168.86.50, server: localhost, request: "POST /mpdplaylist.php HTTP/1.1", upstream: "fastcgi://unix:/run/php-fpm/php-fpm.sock:", host: "raudio-1", referrer: "http://raudio-1/"

**Reply by rern** [reply]
Please try the UPDATE.

**Comment by ceejayemm** [comment]
@rern 
Applied update 20231210 and then applied Update branch at 16:45 GMT.  I am afraid there is no obvious difference after these updates other than the reference to

`File "/usr/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 497, in aiter`

is now

`File "/usr/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 497, in __aiter__`

? This may however be due to the way in which double underscores are shown in Markup ?

I have also reboot the RPi and no other observable difference.

Chris

**Comment by ceejayemm** [comment]
@rern 
Just updated to 20231216 and this is still a problem.  It seems any (all?) changes to the PlayList trigger this set of  errors.

Chris

**Reply by rern** [reply]
Hi Chris,

Prior to latest `python-websockets`, I've seen errors about `websocket` too.

Try: `pacman -Sy python-websockets` > Reboot.

**Comment by ceejayemm** [comment]
@rern 
Thanks.  Updating using the command above to python-websockets 12.x seems to fix the problem.  There is however another small problem to fix in this area:

Dec 16 15:50:58 rAudio-1 nginx[444]: 2023/12/16 15:50:58 [error] 444#444: *28 FastCGI sent in stderr: "PHP message: PHP Warning:  Undefined variable $list in /srv/http/mpdplaylist.php on line 55; PHP message: PHP Warning:  Trying to access array offset on value of type null in /srv/http/mpdplaylist.php on line 55" while reading response header from upstream, client: 192.168.86.53, server: localhost, request: "POST /mpdplaylist.php HTTP/1.1", upstream: "fastcgi://unix:/run/php-fpm/php-fpm.sock:", host: "raudio-1", referrer: "http://raudio-1/"

Changing the line 55 in the file /srv/http/mpdplaylist.php 
from 
        if ( substr( $list[ 3 ], 0, 4 ) === 'cdda' ) {

to
        if ( substr( $lists[ 3 ], 0, 4 ) === 'cdda' ) {

fixes this too.

Many thanks

Chris

**Comment by ceejayemm** [comment]
As per #1647 this seems to be resolved with 20240315.

Chris
